docs/architecture.md

```diff
@@ -0,0 +1,292 @@
+Kubernetes, and now Armada. Confusingly, cluster managers are also sometimes
+referred to as "resource-managers" within the Spark codebase.
+
+### Defining a new Cluster Manager
+
+To define a new cluster manager within Spark one would inherit the `trait`
+[`ExternalClusterManager`](https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/scheduler/ExternalClusterManager.scala)
+and implement the required methods. Mostly they center around the creation,
+management, and deletion of Spark Executors tied to a particular Driver program.
+
+### Integrating with `spark-submit`
+
+While Spark seems to take some care around providing a defined interface for
+individual Cluster Managers, there's still a fair bit of cluster-specific code
+within `SparkSubmit`. Absent upstream integration, `armada-spark` maintains and
+provides it's own `ArmadaSparkSubmit` implementation which serves as a drop-in
+replacement.
+
+Currently the following versions of Spark are supported:
+- 3.3
+- 3.5
+- 4.1
+
+### Using Armada as a "cluster"
+
+While Armada can be treated as a cluster for the purposes of Spark, some care
+and consideration must be taken in order to meet most, if not all, of Spark's
+expectations.
+
+### Driver and Executor (and Cluster Manager) Network Communications
+
+One major requirement that **must** be met is network communication between
+Drivers and Executors. This allows Drivers to submit tasks to the Executors
+and collate their results. `armada-spark` achieves this through the use of
+[kubernetes labels](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/).
+Used correctly, these labels induce the Armada scheduler to schedule Drivers
+and their accompanying Executors to the same kubernetes cluster managed by
+Armada. Kubernetes cluster networking topology guarantees that nodes within the
+cluster will be able to communicate. Scheduling to the same cluster also has
+the benefit of increasing the probability of compute resources being
+"clustered" closely together. Perhaps servers on the same rack, or even a single
+server.
+
+If you refer back to the ["Cluster Manager Overview"](https://spark.apache.org/docs/latest/cluster-overview.html),
+the first image you'll find is a diagram showing a complete, directed graph between the:
+
+- Driver Program
+- Worker Node / Executors
+- Cluster Manager
+
+and while the diagram appears to separate the "Cluster Manager" out as a separate
+program or node, in reality it tends to "live" in the same process as the
+Driver. So by ensuring connectivity between the Driver and Executors, we also
+get connectivity with the "Cluster Manager".
+
+### Driver and Executors as Armada Jobs
+
+Armada, being a multi-cluster Kubernetes manager, takes on the role of
+scheduling "jobs" to the clusters it oversees. For now, it makes the most sense
+to separate each Driver and Executor into their own Armada job. There are 
```

Comment:
is this still true?

Reply:
Yes, and it's a requirement/limitation of Armada's. You can't submit more than one Pod per job.

Reply:
Ah sorry I was thinking of the jobset, not the job.